File: Lab-6/subwords1.cpp
/*
    FUNCTION SUBWORDS(S):
        #Set for avoiding repeated subwords.
        Let W be an empty set(of strings)
        for i = 1 up to |S| do
            for j = i up to |S| do
                insert S[i:j] into W
        Let L be an empty list(of strings)
        for all w € W do:
            append w into L
        sort L
        return L
*/
#include <string>
#include <iostream>
#include <set>
#include <vector>
#include <algorithm>
using namespace std;

vector<string> subwords( string& S ){
    set<string> W;
    for ( int i = 0; i < S.length(); ++i )
        for ( int j=i; j < S.length(); ++j )
            W.insert( S.substr( i, j - i + 1 ) );
    
    vector<string> L;
    
    for ( auto w : W ) L.push_back( w );

    sort( L.begin(), L.end() );
    
    return L;
    
}

int main(){

    string S;

    getline(cin,S);

    vector<string> L = subwords(S);

    for ( auto l : L ) cout << l << endl;

    return 0 ; 
}

File: Lab-6/subwords1.py
from easyinput import read_line
'''
    FUNCTION SUBWORDS(S):
        #Set for avoiding repeated subwords.
        Let W be an empty set(of strings)
        for i = 1 up to |S| do
            for j = i up to |S| do
                insert S[i:j] into W
        Let L be an empty list(of strings)
        for all w € W do:
            append w into L
        sort L
        return L
'''
def subwords(S):
    W = set()
    n = len(S)
    for i in range( n ):
        for j in range( i, n ): W.add(S[i:j+1])
    L = []
    for w in W: L.append(w)
    return sorted(L)
S = read_line()
[print(w) for w in subwords(S)]
